c++/src/kj/compat/http.c++

```diff
@@ -2857,6 +2875,9 @@ private:
 
   kj::Own<WebSocket> ownState;
 
+  uint64_t sentBytes = 0;
```

Comment:
So, `WebSocketPipeImpl` is a little weird. Normally the `WebSocket` class represents one end of a WebSocket. Messages written to it are received by some other `WebSocket` instance somewhere else. But `WebSocketPipeImpl` is actually a reflector -- messages you send to it with `send()` are then received by calling `receive()` _on the same object_.

When you create a WebSocket pipe, under the hood, two `WebSocketPipeImpl`s are constructed to represent messages flowing in the two directions. And then, as a layer on top of that, two `WebSocketPipeEnd`s are constructed to represent the two ends of the pipe. Each end holds references to both underlying `WebSocketPipeImpl`, one to satisfy sends and one for receives. The other end holds the same references in the opposite order.

What all this means is that `WebSocketPipeImpl` doesn't need to have separate `sentBytes` and `receivedBytes`, because the bytes it sends are exactly the same as the bytes it receives. It can just have a single `transferredBytes`.

Moreover, I think the right way to implement this count is not by editing all the individual state types like `BlockedSend` and `BlockedReceive`, but instead modify the methods of `WebSocketPipeImpl` itself. Note that `send()` and `tryPumpFrom()` both "sending" operations, while `receive()` and `pumpTo()` are both "receiving" operations; you should choose either the sending side or the receiving side to do the counting (but not both). Then you can revert all your changes to the `BlockedWhatever` state classes as they don't need to do any counting of their own.

For pump operations, I think the easiest way to determine how many bytes were pumped is to ask the other WebSocket for its counts before and after the operation, and take the difference. E.g. if you count in `pumpTo()`, check `other.sentBytesCount()` before and after the call (since `pumpTo()` receives from this object and sends to `other`).

Note that for `send()`/`receive()`, you should only count the bytes after the operation completes (without being canceled or erroring out), since otherwise no bytes were actually transferred. But for `pumpTo()`, a partial pump is possible. The strategy mentioned above of checking the other WebSocket's counts before and after should therefore be applied whether the pump succeeds, fails, or is canceled. To cover all three cases, use an attached defer, like:

    auto before = other.receivedBytesCount();
    return pumpTo(other).attach(kj::defer([this,before]() {
      transferredBytes += other.receivedBytesCount() - before;
    }));

The attached-defer will be executed on success, failure, and cancellation.

Reply:
I appreciate this explanation, I was getting turned around in this part. Do you think it makes sense for both `sentBytesCount()` and `receivedBytesCount()` to return the `transferredBytes`?

Reply:
That would be fine, or you could have them throw and define a new method. All that matters is that `WebSocketPipeEnd` knows which method to call, since it is the only thing that gets direct access to the underlying `WebSocketPipeImpl`s.

Reply:
We need to be able to call `receivedBytesCount()` on `other` so I have both methods returning `transferredBytes` instead of adding a new function.